File: COMP 314/1. QuickSort/4.quicksort.cpp
#include <iostream>

using namespace std;

void swap(int elements[], int x, int y) {
	int temp = elements[x];
	elements[x] = elements[y];
	elements[y] = temp;
}

int settlePivotElement (int elements[], int lowerRange, int upperRange) {
	int pivotElement = elements[lowerRange];
	int pL = lowerRange + 1;
	int pR = upperRange;
	while ( true ) {
		while(pL <= pR && elements[pL] < pivotElement) { pL += 1; }
		while(pL <= pR && elements[pR] > pivotElement) { pR -= 1; }
		if (pL > pR) {
			break;
		}
		swap(elements, pL, pR);
	}

	swap(elements, lowerRange, pR);
	return pR;
}

void quickSort (int elements[], int lowerRange, int upperRange ) {
	int indexOfPivotElement = settlePivotElement(elements, lowerRange, upperRange);

	if (lowerRange <= indexOfPivotElement - 1 ) {
		quickSort(elements, lowerRange, indexOfPivotElement - 1);
	}

	if (indexOfPivotElement + 1 <= upperRange) {
		quickSort(elements, indexOfPivotElement + 1 , upperRange);
	}
}

int main() {
	int input[] = { 11, 32, 99, 5, 206, 258, 111, 25, 252, 82, 434,23, 477, 1111, 1, 24,362,364,250,177,295,241,411,441,234,263,386,47,222,368,260,374,291,404,74};
	int inputSize = (sizeof(input)/sizeof(*input));
	cout << "Input Size = " << inputSize << endl;
	quickSort(input, 0, inputSize);
	for (int i = 0; i < inputSize; i++ ) {
		cout << input[i] << " ";
	}
	cout << endl;
}

File: Year III/COMP 314/1. QuickSort/4.quicksort.py
# Implementation of algorithm as discussed in class

def swap(arr, x, y):
	(arr[x], arr[y]) = (arr[y], arr[x])

def settlePivotElement(elements, lower_range, upper_range):
	pivotElement = elements[lower_range]
	left_pointer = lower_range + 1
	right_pointer = upper_range
	while True:
		while left_pointer <= right_pointer and elements[left_pointer] < pivotElement:
			left_pointer += 1
		while left_pointer <= right_pointer and elements[right_pointer] > pivotElement: 
			right_pointer -= 1
		if left_pointer > right_pointer: break
    
		swap(elements, left_pointer, right_pointer)

	swap(elements, lower_range, right_pointer)
	return right_pointer

def quickSort(elements, lower_range, upper_range):
	index_of_pivot_element = settlePivotElement(elements, lower_range, upper_range)
	
	if lower_range <= index_of_pivot_element - 1:
		quickSort(elements, lower_range, index_of_pivot_element - 1)

	if index_of_pivot_element + 1 <= upper_range:
		quickSort(elements, index_of_pivot_element + 1 , upper_range)

if __name__ == "__main__":
	inputs = [11, 32, 99, 5, 206, 258, 111, 25, 252, 82, 434, 23, 477, 1111, 1]
	print("Unsorted Input :: ", inputs)
	lower_range = 0
	upper_range = len(inputs) - 1
	quickSort(inputs, lower_range, upper_range)
	print("Sorted Output  :: ",inputs)
